SeQuant/core/index.hpp

```diff
@@ -79,62 +110,173 @@ class Index : public Taggable {
 
   Index() = default;
 
-  /// @param label the label, does not need to be unique
+  /// copy constructor
+  /// @warning memoized data (label, full_label) is not copied
+  Index(const Index &idx) noexcept
+      : Taggable(idx),
+        space_(idx.space_),
+        ordinal_(idx.ordinal_),
+        proto_indices_(idx.proto_indices_),
+        symmetric_proto_indices_(idx.symmetric_proto_indices_) {}
+
+  /// move constructor
+  /// @note memoized data (label, full_label) is copied
+  Index(Index &&) = default;
+
+  /// copy assignment
+  /// @warning memoized data (label, full_label) is not copied
+  Index &operator=(const Index &idx) {
+    Taggable::operator=(idx);
+    space_ = idx.space_;
+    ordinal_ = idx.ordinal_;
+    proto_indices_ = idx.proto_indices_;
+    symmetric_proto_indices_ = idx.symmetric_proto_indices_;
+    return *this;
+  }
+
+  /// move assignment
+  /// @note memoized data (label, full_label) is copied
+  Index &operator=(Index &&) = default;
+
+  /// @param space (a const ref to) the IndexSpace object that specifies to this
+  /// space this object belongs
+  explicit Index(const IndexSpace &space) : Taggable(), space_(space) {
+    check_nonreserved();
+  }
+
+  /// @param space (a const ref to) the IndexSpace object that specifies to this
+  /// space this object belongs
+  /// @param ord the index ordinal
+  Index(const IndexSpace &space, meta::integral auto ord)
+      : space_(space), ordinal_(to_ordinal(ord)) {
+    check_nonreserved();
+  }
+
+  /// @param space (a const ref to) the IndexSpace object that specifies to this
+  /// space this object belongs
+  /// @param ord the index ordinal
+  template <meta::integral I>
+  Index(const IndexSpace &space, std::optional<I> ord)
+      : space_(space), ordinal_(ord) {
+    check_nonreserved();
+  }
+
   /// @param space (a const ref to) the IndexSpace object that specifies to this
   /// space this object belongs
   /// @param proto_indices labels of proto indices (all must be unique,
   /// i.e. duplicates are not allowed)
   /// @param symmetric_proto_indices if true, proto_indices can be permuted at
   /// will and will always be sorted
-  template <typename String,
-            typename = std::enable_if_t<meta::is_basic_string_convertible_v<
-                std::remove_reference_t<String>>>>
-  Index(String &&label, const IndexSpace &space, IndexList proto_indices,
+  Index(const IndexSpace &space, container::vector<Index> proto_indices,
         bool symmetric_proto_indices = true)
-      : label_(to_wstring(std::forward<String>(label))),
-        space_(space),
-        proto_indices_(proto_indices),
-        symmetric_proto_indices_(symmetric_proto_indices) {
+      : Index(space) {
+    proto_indices_ = std::move(proto_indices);
+    symmetric_proto_indices_ = symmetric_proto_indices;
```

Comment:
```suggestion
      : Index(space), proto_indices_ = std::move(proto_indices), symmetric_proto_indices_ = symmetric_proto_indices {
```

Reply:
delegating ctor can't be supplemented by field initialization